design/view-change.md

```diff
@@ -227,151 +227,97 @@ we don't place any upper limit on this time.
    can be applied only to master instance, with backups synced to master
    during view change.
 
-## Rough roadmap
-
-1. Implement PBFT view change as pluggable strategy that can cope with
-   resetting ppSeqNo. This will be simplest possible variant with disabled
-   request processing during view change, dropped requests that come after
-   gaps and so on. Check that this implementation works. If it does,
-   and performs better than current view change go ahead to we can take
-   faster route A, otherwise move to route B.
-
-Route A:
-
-1. Set new view change as default, remove implementation of old view change
-   and all related code, which should simplify next changes.
-
-2. Fix current code so that it doesn't rely on resetting ppSeqNo after
-   view change.
-
-3. Implement other pieces of PBFT view change, which improve performance
-   and/or pool responsiveness during view change.
-
-Route B:
+## Proposed deviations
+
+1. Reset pp_seq_no to 1 with a new view
+2. Order 3PC Batches from the last view (with a NEW_VIEW msg) with the previous view's viewNo. This is needed to have consistent audit ledger.
+
+
+#### Plan of Attack
+
+- NEW_VIEW needs to consist of a list of PrePrepares (with all the information such as txn and state roots). NEW_VIEW is not passed to Orderer, only the PrePrepares are passed.
+- Reset uncommitted state once start the view change
+- Introduce a new `wait_for_new_view` flag to distinguish old-way and new-way view changes
+
+    - This will be a flag in a ConsensusDataProvider
+    - The flag is set to TRUE once view change is started (and viewNo +=1)
+    - The flag is set to FALSE once a valid NEW_VIEW is received
+
+- There needs to be `last_prep_cert` field in ConsensusDataProvider: (last_view_no, prep_cert_pp_seq_no)
```

Comment:
What would happen if there are two view changes in a row, with some old batches still not finished and some new batches slipping in between?

Reply:
How new batches can slip in between if we require a strict order of PrePrepares and apply them in this order only? Also we don't apply a NEW_VIEW (and the corresponding PrePrepares) if NEW_VIEW is not valid (doesn't have all PrePrepares for the `last_prep_cert`).